o2a/o2a_libs/el_parser.py

```diff
@@ -0,0 +1,296 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains tools for translating Expression Language to Jinja.
+
+The BNF for the grammar is based on official grammar of EL:
+https://download.oracle.com/otn-pub/jcp/jsp-2.1-fr-spec-oth-JSpec/jsp-2_1-fr-spec-el.pdf
+"""
+
+__all__ = ["translate"]
+
+from typing import Union
+import re
+
+from lark import Lark, Tree, Token
+
+
+GRAMMAR = r"""
+    start: (lvalue (start)?)* | (rvalue (start)?)* | literal_expression (rvalue (start)?)?
+
+    lvalue: "${" lvalue_inner "}" | "#{" lvalue_inner "}"
+
+    rvalue: ("${" expression "}")+ | ("#{" expression "}")+
+
+    lvalue_inner: identifier
+        | non_literal_lvalue_prefix (value_suffix)*
+
+    literal_expression: literal_component (/[\$\#]/)?
+
+    // TODO: extend to full rule ! ! !
+    literal_component: (/[^\$\#]/)*
+        | (/[^\$\#]/)* (/[\$\#][^\{]/)
+        | (/[^\$\#]/)* /\\\\/ (/[\$\#]/)?
+
+    expression: expression1 ternary?
+
+    ternary: "?" expression ":" expression
+
+    expression1: expression binary_op expression
+        | unary_expression
+
+    binary_op: "and"
+        | "&&"
+        | "or"
+        | "||"
+        | "+"
+        | "-"
+        | "*"
+        | "/"
+        | "div"
+        | "%"
+        | "mod"
+        | ">"
+        | "gt"
+        | "<"
+        | "lt"
+        | ">="
+        | "ge"
+        | "<="
+        | "le"
+        | "=="
+        | "eq"
+        | "!="
+        | "ne"
+
+    unary_expression: unary_op unary_expression
+        | value
+
+    unary_op: "-"
+        | "!"
+        | "not"
+        | "empty"
+
+    value: value_prefix (value_suffix)*
+
+    value_prefix: literal
+        | non_literal_lvalue_prefix
+
+    non_literal_lvalue_prefix: "(" expression ")"
+        | identifier
+        | function_invocation
+
+    value_suffix: "." identifier | "[" expression "]"
+
+    identifier: JAVA
+
+    function_invocation: (identifier ":")? identifier "(" ( expression ( "," expression )* )? ")"
+
+    literal: BOOL | INT | FLOAT | STRING | NULL
+
+    JAVA: /[a-zA-Z_]+/
+
+    BOOL: "true" | "false"
+
+    STRING: /\'[^\']*\'/ | /\"[^\"]*\"/
+
+    INT: /0|[1-9]\d*/i
+
+    FLOAT: (/[0-9]/)+ "." (/[0-9]/)* EXP?
+        | "." (/[0-9]/)+ EXP?
+        | (/[0-9]/)+ EXP?
+
+    EXP: /[eE]/ (/[\+\-]/)? (/[0-9]/)+
+
+    NULL: "null"
+
+    %ignore " "
+"""
+
+
+def _parser(sentence: str) -> Tree:
+    return Lark(GRAMMAR, start="start", keep_all_tokens=True).parse(sentence)
+
+
+def _camel_to_snake(name: str) -> str:
```

Comment:
are there standard py libs for this? seems like a common need.

Reply:
It seems that using regex is the most popular approach and probably the simplest.

Reply:
We all got curious about it but it seems this is THE most popular solution :) with 700 upvotes on Stack Overflow